llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp

```diff
@@ -1244,17 +1247,44 @@ InstructionCost SystemZTTIImpl::getInterleavedMemoryOpCost(
   return NumVectorMemOps + NumPermutes;
 }
 
-static int getVectorIntrinsicInstrCost(Intrinsic::ID ID, Type *RetTy) {
+static int getVectorIntrinsicInstrCost(
+    Intrinsic::ID ID,
+    Type *RetTy,
+    const SmallVectorImpl<Type *>& ParamTys) {
   if (RetTy->isVectorTy() && ID == Intrinsic::bswap)
     return getNumVectorRegs(RetTy); // VPERM
+  else if (ID == Intrinsic::vector_reduce_add) {
+    // Retrieve number and size of elements for the vector op
+    auto *VTy = cast<FixedVectorType>(ParamTys.front());
+    unsigned NumElements = VTy->getNumElements();
+    unsigned BitWidth = VTy->getScalarSizeInBits();
+    // For scalar sizes <128 bits, we fall back to the generic cost estimate
+    if (BitWidth > SystemZ::VectorBits)
+      return -1;
+    // How many elements can a single vector register hold
+    unsigned MaxElemsPerVector = SystemZ::VectorBits / BitWidth;
+    // How many vector regs are needed to represent the input elements (V)
+    unsigned VectorRegsNeeded = getNumVectorRegs(VTy);
+    assert(VectorRegsNeeded > 0 && "Vector register count must not be 0");
```

Comment:
There is already an assertion in getNumVectorRegs(). I think this really should always be at least one vector reg for any vector type, so in case you want to add the assertion, maybe do it inside getNumVectorRegs()?

Reply:
Thanks for pointing that out, i had missed that. I was mainly concerned that if there is some problem with the computation, and i end up with the final cost evaluating to something like -1, that the system would silently use the generic cost computation instead. I have changed the assert to something closer to that intent, do you think that's better? Or would you say that errors like this would be caught by the costmodel tests anyway, and there's no need to be concerned?
I mostly put the asserts there because the style guide you linked suggested to spread them liberally across the code base ;)

Reply:
I like the new assertion more: This is more generic and it checks the final result which is always good, and it's easy to read. If this code is changed in the future the assert would probably be good to have there in place, even though at the moment it is trivial to see that it could never be zero. For me it's kind of a border case, so I have no opinion on having it there or not. As you say, with your new tests that would be kind of a very weird problem to return 0 in some untested case...

Reply:
ok, then on balance let's say we keep the assert where it is for now.